### PR TITLE
[Storage Blobs Service] Transactional MD5 and CRC64

### LINKED ISSUE #295: [Storage Blob Service] Transactional Checksum Support in storage SDK
MD5 and CRC64

### PR DESCRIPTION
**Pull request by JinmingHu-MSFT**
closes #295

### CODE REVIEW

**Comment by JinmingHu-MSFT** [comment]
closes #295

File: sdk/storage/inc/common/crypt.hpp [hunk 1]
```diff
@@ -3,13 +3,65 @@
...
+  class Md5 {
+  public:
+    Md5();
+    ~Md5();
+
+    void Update(const uint8_t* data, std::size_t length);
+
+    std::string Digest() const;
+
+    static std::string Hash(const uint8_t* data, std::size_t length)
```
Comment (katmsft) [inline sdk/storage/inc/common/crypt.hpp:23]: A one-liner saying that this API returns Base64 encoded string would be great.

File: sdk/storage/src/common/crypt.cpp [hunk 2]
```diff
@@ -82,96 +97,164 @@ namespace Azure { namespace Storage {
...
   {
+    std::string buffer;
+    BCRYPT_HASH_HANDLE hashHandle = nullptr;
+    std::size_t hashLength = 0;
+  };
 
-    static AlgorithmProviderInstance AlgorithmProvider(AlgorithmType::Hmac_Sha256);
+  Md5::Md5()
+  {
+    static Details::AlgorithmProviderInstance AlgorithmProvider(Details::AlgorithmType::Md5);
```
Comment (katmsft) [inline sdk/storage/src/common/crypt.cpp:201]: Why pascal case?

Comment (katmsft) [reply sdk/storage/src/common/crypt.cpp:201]: Ditto for other local variables. I see snake case/pascal case/camel cases in different areas.

File: sdk/storage/inc/common/crypt.hpp (hunk 1 above)
Comment (JinmingHu-MSFT) [reply sdk/storage/inc/common/crypt.hpp:23]: It returns binary

File: sdk/storage/src/common/crypt.cpp [hunk 3]
```diff
@@ -275,4 +389,823 @@ namespace Azure { namespace Storage {
 
 #endif
 
+  static constexpr uint64_t poly = 0x9A6C9329AC4BC9B5ULL;
```
Comment (katmsft) [inline sdk/storage/src/common/crypt.cpp:392]: Good thing is that track 1 proved the below code's integrity. Otherwise, I don't know how to maintain it.

Comment (JinmingHu-MSFT) [reply sdk/storage/src/common/crypt.cpp:392]: abstruse mathematical principles behind it. Don't try to comprehend it. just use it.

File: sdk/storage/src/common/crypt.cpp (hunk 2 above)
Comment (JinmingHu-MSFT) [reply sdk/storage/src/common/crypt.cpp:201]: fixed

**Pull request merged**